File: examples/boschProcess/boschProcessEmulate.py
from argparse import ArgumentParser
import viennaps as ps

# parse config file name and simulation dimension
parser = ArgumentParser(
    prog="boschProcessEmulate",
    description="Run a Bosch process emulation on a trench geometry.",
)
parser.add_argument("-D", "-DIM", dest="dim", type=int, default=2)
parser.add_argument("filename")
args = parser.parse_args()

# switch between 2D and 3D mode
if args.dim == 2:
    print("Running 2D simulation.")
else:
    print("Running 3D simulation.")
ps.setDimension(args.dim)

ps.Logger.setLogLevel(ps.LogLevel.ERROR)
params = ps.readConfigFile(args.filename)
ps.setNumThreads(16)

geometry = ps.Domain(
    gridDelta=params["gridDelta"],
    xExtent=params["xExtent"],
    yExtent=params["yExtent"],
)
ps.MakeTrench(
    domain=geometry,
    trenchWidth=params["trenchWidth"],
    trenchDepth=0.0,
    maskHeight=params["maskHeight"],
).apply()


direction = [0.0, 0.0, 0.0]
direction[args.dim - 1] = -1.0

# Geometric advection model for deposition
depoModel = ps.SphereDistribution(
    radius=params["depositionThickness"], gridDelta=params["gridDelta"]
)

# Define purely directional rate for depo removal
etchDir = ps.RateSet(
    direction=direction,
    directionalVelocity=-(params["depositionThickness"] + params["gridDelta"] / 2.0),
    isotropicVelocity=0.0,
    maskMaterials=[ps.Material.Mask],
)
depoRemoval = ps.DirectionalProcess(rateSets=[etchDir])

# Define isotropic + direction rate for etching of substrate
etchIso = ps.RateSet(
    direction=direction,
    directionalVelocity=params["ionRate"],
    isotropicVelocity=params["neutralRate"],
    maskMaterials=[ps.Material.Mask, ps.Material.Polymer],
)
etchModel = ps.DirectionalProcess(rateSets=[etchIso])
etchTime = params["etchTime"]

n = 0


def runProcess(model, name, time=1.0):
    global n
    print("  - {} - ".format(name))
    ps.Process(geometry, model, time).apply()
    geometry.saveSurfaceMesh("boschProcessEmulate_{}".format(n))
    n += 1


def cleanup(threshold=1.0):
    expand = ps.IsotropicProcess(threshold)
    ps.Process(geometry, expand, 1).apply()
    shrink = ps.IsotropicProcess(-threshold)
    ps.Process(geometry, shrink, 1).apply()


numCycles = int(params["numCycles"])

# Initial geometry
geometry.saveSurfaceMesh("boschProcessEmulate_{}".format(n))
n += 1

runProcess(etchModel, "Etching", etchTime)

for i in range(numCycles):
    print("Cycle {}".format(i + 1))

    # Deposit a layer of polymer
    geometry.duplicateTopLevelSet(ps.Material.Polymer)
    runProcess(depoModel, "Deposition")

    # Remove the polymer layer
    runProcess(depoRemoval, "Punching through")

    # Etch the trench
    runProcess(etchModel, "Etching", etchTime)

    # Ash (remove) the polymer
    geometry.removeTopLevelSet()
    cleanup(params["gridDelta"])
    geometry.saveSurfaceMesh("boschProcessEmulate_{}".format(n))
    n += 1

# save the final geometry
geometry.saveVolumeMesh("boschProcessEmulate_final")

File: examples/boschProcess/boschProcessEmulate.cpp
#include <geometries/psMakeTrench.hpp>
#include <models/psDirectionalProcess.hpp>
#include <models/psGeometricDistributionModels.hpp>
#include <models/psIsotropicProcess.hpp>
#include <process/psProcess.hpp>
#include <psUtil.hpp>

using namespace viennaps;
constexpr int D = 2;
using NumericType = double;

void etch(SmartPointer<Domain<NumericType, D>> &domain,
          util::Parameters &params) {
  std::cout << "  - Etching - " << std::endl;
  DirectionalProcess<NumericType, D>::RateSet rateSet;
  rateSet.direction = Vec3D<NumericType>{0.};
  rateSet.direction[D - 1] = -1.;
  rateSet.directionalVelocity = params.get("ionRate");
  rateSet.isotropicVelocity = params.get("neutralRate");
  rateSet.maskMaterials =
      std::vector<Material>{Material::Mask, Material::Polymer};
  rateSet.calculateVisibility = true;

  auto etchModel =
      SmartPointer<DirectionalProcess<NumericType, D>>::New(rateSet);
  Process<NumericType, D>(domain, etchModel, params.get("etchTime")).apply();
}

void punchThrough(SmartPointer<Domain<NumericType, D>> &domain,
                  util::Parameters &params) {
  std::cout << "  - Punching through - " << std::endl;
  DirectionalProcess<NumericType, D>::RateSet rateSet;
  rateSet.direction = Vec3D<NumericType>{0.};
  rateSet.direction[D - 1] = -1.;
  rateSet.directionalVelocity =
      -(params.get("depositionThickness") + params.get("gridDelta") / 2.);
  rateSet.isotropicVelocity = 0.;
  rateSet.maskMaterials = std::vector<Material>{Material::Mask};
  rateSet.calculateVisibility = true;

  // punch through step
  auto depoRemoval =
      SmartPointer<DirectionalProcess<NumericType, D>>::New(rateSet);
  Process<NumericType, D>(domain, depoRemoval, 1.).apply();
}

void deposit(SmartPointer<Domain<NumericType, D>> &domain,
             NumericType depositionThickness) {
  std::cout << "  - Deposition - " << std::endl;
  domain->duplicateTopLevelSet(Material::Polymer);
  auto model = SmartPointer<SphereDistribution<NumericType, D>>::New(
      depositionThickness);
  Process<NumericType, D>(domain, model).apply();
}

void ash(SmartPointer<Domain<NumericType, D>> &domain) {
  domain->removeTopLevelSet();
  domain->removeStrayPoints();
}

int main(int argc, char **argv) {

  Logger::setLogLevel(LogLevel::ERROR);
  omp_set_num_threads(16);

  // Parse the parameters
  util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
    return 1;
  }

  // geometry setup
  auto geometry = Domain<NumericType, D>::New(
      params.get("gridDelta"), params.get("xExtent"), params.get("yExtent"));
  MakeTrench<NumericType, D>(geometry, params.get("trenchWidth"),
                             0.0, // trenchDepth
                             0.0, // trenchTaperAngle
                             params.get("maskHeight"))
      .apply();

  const NumericType depositionThickness = params.get("depositionThickness");
  const int numCycles = params.get<int>("numCycles");
  const std::string name = "boschProcessEmulate_";

  int n = 0;
  geometry->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
  etch(geometry, params);
  geometry->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");

  for (int i = 0; i < numCycles; ++i) {
    std::cout << "Cycle " << i + 1 << std::endl;
    deposit(geometry, depositionThickness);
    geometry->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
    punchThrough(geometry, params);
    geometry->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
    etch(geometry, params);
    geometry->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
    ash(geometry);
    geometry->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
  }

  geometry->saveVolumeMesh("boschProcessEmulate_final");
}